src/guide/migration/attrs-includes-class-style.md

```diff
@@ -0,0 +1,69 @@
+---
+title: $attrs includes class & style
+badges:
+  - breaking
+---
+
+# `$attrs`includes `class` & `style` <MigrationBadges :badges="$frontmatter.badges" />
+
+## Overview
+
+`$attrs` now contains _all_ attributes passed to a component, including `class` and `style.`
+
+## 2.x Behavior
+
+`class` and `style` attributes get some special handling in the Vue 2 VirtualDOM implementation. For that reason, they are _not_ included in `$attrs`, while all other attributes are.
+
+A side effect of this manifests when using `inheritAttrs: false`:
+
+- Attributes in `$attrs` are no longer automatically added to the root element, leaving it the developer to decide where to add them.
+- But `class` and `style`, not being part of `$attrs`, will still be applied to the component's root element:
+
+```html
+<template>
+  <label>
+    <input type="text" v-bind="$attrs" />
+  </label>
+</template>
+<script>
+  export default {
+    inheritAttrs: false
+  }
+</script>
+```
+
+when used like this:
+
+```html
+<my-component id="my-id" class="my-class"></my-component>
+```
+
+...will generate this HTML:
+
+```html
+<label class="my-class">
+  <input type="text" id="my-id" />
+</label>
+```
+
+## 3.x Behavior
+
+`$attrs` contains _all_ attributes, which makes it easier to apply all of them to a different element. The example from above now generates the following HTML:
+
+```html
+<label>
+  <input type="text" id="my-id" class="my-class" />
+</label>
+```
+
+## Migration Strategy
```

Comment:
`class` is commonly used for layout. For that purpose, it is usually necessary for the classes to be applied to the outer element, not the inner element. It isn't really clear from this migration section how to make that work.

Personally I've been defining `class` and `style` in `props` so they won't be in `$attrs` and can be applied to the outer element.

I've not seen anything to suggest what the 'official' approach is on this.

Reply:
> I've not seen anything to suggest what the 'official' approach is on this.

There's usually is not *one* official approach, as for most things as flexible as `$attrs`.

Instead of defining props you could compute a newe object from attr that omits class and style a then apply `$attrs.class` in one place and `restOfAttrs` in another. I don't see an argument for one over the other.

I'm willing to add more examples here but would need some realistic input. I'm probably too close to the metal to have a feeling what users may be unsure about here.